### PR TITLE
Implement hasChanged on Slots (closes #311 #312 #310)

### LINKED ISSUE #311: Remove redundant signal propagation
The invalidation signal should propagate through the pipeline.
However, if it passes an already invalid slot, propagation should stop to boost performance.

### PR DESCRIPTION
**Pull request by chistopher**
See #311 #312 #310

### CODE REVIEW

File: source/gloperate/include/gloperate/pipeline/Slot.inl [hunk 1]
```diff
@@ -183,10 +188,26 @@ void Slot<T>::setValid(bool isValid)
     }
 
     // Set state of own data
+    auto wasValid = m_valid;
     m_valid = isValid;
 
-    // Emit signal
-    this->onValueChanged(this->m_value);
+    // Emit signal if it was invalidated
+    if(wasValid && !isValid)
```
Comment (chistopher) [inline source/gloperate/include/gloperate/pipeline/Slot.inl]: Does this conform to #310. Invalidation is signaled (including propagation) when the slot changes from valid to invalid; or when the connected source invalidates (see Slot::connect). That means setting a Slot to valid can never trigger the signal nor can an invalid Slot signal again.

File: source/gloperate/include/gloperate/pipeline/Output.inl [hunk 2]
```diff
@@ -46,37 +46,18 @@ void Output<T>::onRequiredChanged()
...
-    if (this->m_cycleGuard[this_id])
-    {
-        // Reset guard
-        this->m_cycleGuard[this_id] = false;
-
-        // Stop recursion here to avoid endless recursion
-        return;
-    }
+    // Emit signal
+    this->valueInvalidated();
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Output.inl]: Why doesn't an output require cycle detection anymore?

File: source/gloperate/include/gloperate/pipeline/Slot.inl [hunk 3]
```diff
@@ -76,11 +76,15 @@ bool Slot<T>::connect(Slot<T> * source)
     // Set source
     m_source = source;
 
-    // Connect to data container
-    m_connection = m_source->valueChanged.connect([this] (const T & value)
+    // Connect to data container; no direct binding of member function to achive virtual lookup
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Slot.inl]: :+1: valuable comment

File: source/gloperate/include/gloperate/pipeline/Slot.inl [hunk 4]
```diff
@@ -183,10 +188,14 @@ void Slot<T>::setValid(bool isValid)
     }
 
     // Set state of own data
+    auto wasValid = m_valid;
```
Comment (scheibel) [inline source/gloperate/include/gloperate/pipeline/Slot.inl]: I think it should be a ```const auto``` type.

**Pull request merged**